### PR TITLE
Refactor `Receipt` deserialization

### LINKED ISSUE #12400: refactor BorshDeserialization for Receipt
The deserialization for receipt is a custom implementation that first checks the first few bytes and based on that chooses to deserialize the receipt as either V0 or V1. Since the first bytes are read from the Reader, each of the V0/V1 deserializations is done field by field. It's quite ugly. The goal for this issue is to refactor the implementation so that the deserialization of V0 and V1 can be used directly. It can be done using the following trick:
- read the necessary bytes to decide how to deserialize
- prepare a buffer that consists of the read bytes and the remaining reader - it can be done by using `chain`
  - e.g. https://github.com/near/nearcore/blob/864270f15964e0fa0e8fb1ddc4ec73c358f0bac9/core/primitives/src/receipt.rs#L329
- deserialize using the V0/V1 deserialization

Please note that the actual data layout should not change - it's only the implementation of the deserialization that we're trying to make nicer.

### PR DESCRIPTION
**Pull request by eagr**
closes #12400

### CODE REVIEW

**Review by tayfunelmas — APPROVED** [review]

**Comment by tayfunelmas** [comment]
Can you check if there are enough tests for this change (eg. deserialization of V0 and V1) and add any in this PR?

**Comment by eagr** [comment]
The happy path is covered, but not the error cases, which would occur with tampered or corrupted packets. Do you think it's good enough?

File: core/primitives/src/receipt.rs [hunk 1]
```diff
@@ -210,49 +210,30 @@ impl BorshDeserialize for Receipt {
...
-            })?)
-            .map_err(|e| Error::new(ErrorKind::InvalidData, e.to_string()))
-        };
+        let u1 = u8::deserialize_reader(reader)?;
+        let u2 = u8::deserialize_reader(reader)?;
+
+        let prefix = if u2 == 0 { vec![u1, u2] } else { vec![u2] };
+        let mut reader = prefix.chain(reader);
+
+        let predecessor_id = AccountId::deserialize_reader(&mut reader)?;
```
[the remark this answers is not in the document]
Comment (eagr) [reply core/primitives/src/receipt.rs:226]: oh yeah, that's much better

File: core/primitives/src/receipt.rs [hunk 2]
```diff
@@ -210,55 +210,25 @@ impl BorshDeserialize for Receipt {
...
-            }
-            AccountId::try_from(String::from_utf8(str_vec).map_err(|_| {
-                Error::new(ErrorKind::InvalidData, "Failed to parse AccountId from bytes")
-            })?)
-            .map_err(|e| Error::new(ErrorKind::InvalidData, e.to_string()))
-        };
+        let u1 = u8::deserialize_reader(reader)?;
+        let u2 = u8::deserialize_reader(reader)?;
+
+        let prefix = if u2 == 0 { vec![u1, u2] } else { vec![u2] };
```
Comment (wacban) [inline core/primitives/src/receipt.rs:223]: mini nit: I slightly dislike that this check (u2 == 0) happens in two places. Maybe consider moving it inside the respective if cases? Or make it into a variable `let is_v0 = u2 == 0`?

**Review by wacban — APPROVED** [review]
LGTM, thank you, it looks so much better now! 

Regarding tests - I can see there are some simple tests for v0 and v1 serialization. You haven't introduced any new errors or panics and the modified method already returns an error so I'm not too worried about the invalid or malicious cases.

File: core/primitives/src/receipt.rs (hunk 2 above)
Comment (eagr) [reply core/primitives/src/receipt.rs:223]: nice catch. 2nd way is better

**Pull request merged**